commit 852e0943f3880f9a5a4100abb6546e8ab19faf47 (refs/changes/02/1302/2)
Author: Cody Peter Mello <cody.mello@joyent.com>
Date:   2017-01-20T22:22:50+00:00 (2 years, 9 months ago)
    
    FWAPI-272 "icmp6" protocol should be restricted to a minimum of version 3

diff --git a/lib/parser.js b/lib/parser.js
index f50b0ab..23f64d3 100644
--- a/lib/parser.js
+++ b/lib/parser.js
@@ -20,7 +20,7 @@
  *
  * CDDL HEADER END
  *
- * Copyright (c) 2016, Joyent, Inc. All rights reserved.
+ * Copyright 2017, Joyent, Inc. All rights reserved.
  *
  *
  * fwadm: firewall rule parser
@@ -93,7 +93,8 @@ case 33: this.$ = { 'name': $$[$0-1].toLowerCase(), 'targets': $$[$0] }
 break;
 case 34: this.$ = { 'name': $$[$0-1].toLowerCase(), 'targets': $$[$0] } 
 break;
-case 35: this.$ = { 'name': $$[$0-1].toLowerCase(), 'targets': $$[$0] } 
+case 35: yy.validateOKVersion(3, 'IPv6');
+          this.$ = { 'name': $$[$0-1].toLowerCase(), 'targets': $$[$0] }; 
 break;
 case 36: this.$ = $$[$0-1]; 
 break;
diff --git a/src/fwrule.jison b/src/fwrule.jison
index cc6140a..0f48785 100644
--- a/src/fwrule.jison
+++ b/src/fwrule.jison
@@ -20,7 +20,7 @@
  *
  * CDDL HEADER END
  *
- * Copyright (c) 2016, Joyent, Inc. All rights reserved.
+ * Copyright 2017, Joyent, Inc. All rights reserved.
  *
  *
  * fwadm: firewall rule parser grammar
@@ -216,7 +216,8 @@ protocol
     | ICMP type_list
         { $$ = { 'name': $1.toLowerCase(), 'targets': $2 } }
     | ICMP6 type_list
-        { $$ = { 'name': $1.toLowerCase(), 'targets': $2 } }
+        { yy.validateOKVersion(3, 'IPv6');
+          $$ = { 'name': $1.toLowerCase(), 'targets': $2 }; }
     ;
 
 
diff --git a/src/header.js b/src/header.js
index c7e8f7a..947407a 100644
--- a/src/header.js
+++ b/src/header.js
@@ -20,7 +20,7 @@
  *
  * CDDL HEADER END
  *
- * Copyright (c) 2016, Joyent, Inc. All rights reserved.
+ * Copyright 2017, Joyent, Inc. All rights reserved.
  *
  *
  * fwadm: firewall rule parser
diff --git a/test/parser.test.js b/test/parser.test.js
index 55373eb..b278691 100644
--- a/test/parser.test.js
+++ b/test/parser.test.js
@@ -20,7 +20,7 @@
  *
  * CDDL HEADER END
  *
- * Copyright (c) 2016, Joyent, Inc. All rights reserved.
+ * Copyright 2017, Joyent, Inc. All rights reserved.
  *
  *
  * Unit tests for the firewall rule parser
@@ -361,17 +361,31 @@ test('port ranges', function (t) {
 });
 
 
-test('version mismatch', function (t) {
-    try {
-        parser.parse('FROM tag foo TO tag bar ALLOW TCP PORTS 20-30',
-            { maxVersion: 1 });
-        t.ok(false,
-            'Using port ranges is a newer feature and should fail in v1');
-    } catch (err) {
-        t.deepEqual(err.message,
-            'The rule uses a feature (port ranges) newer than this API allows',
-            'Correct error message for using ports in version 1');
-    }
+test('Parser option: maxVersion', function (t) {
+    var versionFmt = 'The rule uses a feature (%s) newer than this API allows';
+
+    [
+        // Version 2 features:
+        [ 'FROM tag foo TO tag bar ALLOW TCP PORTS 20-30', 1, 'port ranges' ],
+
+        // Version 3 features:
+        [ 'FROM tag a to ip fd00::1 ALLOW tcp PORT 80', 2, 'IPv6' ],
+        [ 'FROM tag a to subnet fd00::/64 ALLOW tcp PORT 80', 2, 'IPv6' ],
+        [ 'FROM tag a to tag b ALLOW icmp6 TYPE 135', 2, 'IPv6' ],
+        [ 'FROM tag a to tag b ALLOW icmp TYPE ALL', 2, 'all ICMP types' ]
+    ].forEach(function (cfg) {
+        var rule = cfg[0];
+        var v = cfg[1];
+
+        try {
+            parser.parse(cfg[0], { maxVersion: v });
+            t.fail(util.format('Should fail in v%d: %d', v, rule));
+        } catch (err) {
+            t.deepEqual(err.message, util.format(versionFmt, cfg[2]),
+                util.format('Correct error message when using v%d: ', v, rule));
+        }
+    });
+
     t.end();
 });
 
